### ISSUE TITLE
Simplifying Custom OAuth Login

### ISSUE THREAD

**Issue by taliaga** [rancher/rancher#8193]
Hi there,

We're glad to be using rancher. Moreover, we are interested in supporting user logins with a custom OAuth2 IDM.

We are aware that the first [authentication support](https://github.com/rancher/rancher/issues/20) milestone was for github. Also, it was good to see auth was extracted as a separate service in 1.12. It seems there might be still some [dependencies in the ui](https://github.com/rancher/ui/pull/18/files) though.

Also, it seems there are many people are also interested in supporting this as it can be seen in [Gitlab OAuth](https://github.com/rancher/rancher/issues/1317), [Google SignIn](https://github.com/rancher/rancher/issues/1411) and [Bitbucket Oauth](https://github.com/rancher/rancher/issues/113).

By looking at how the github support was implemented we tested some hacks to make our login work. However, this involved rebuilding a couple of services with some difficulties. But going long term syncing such a hack in a fork might not be that nice.

In search of a better solution, we wanted to know if...

- Are there plans from rancher to go a step further to simplify the implementation of OAuth service X? I'm talking with consideration of [this comment](https://github.com/rancher/rancher/issues/1317#issuecomment-239398843) from @vincent99.
- If not, would you be able to give us some pointers regarding what could be improved to give it a shot?
- If you are reading this issue and actually had to do something similar, what/how did you do?

Final note: We are not even talking about the considerations to support multiple IM providers at the same time. We're just thinking of making it more "simpler / official" to change github to whatever other provider works more or less in the same line. We could work on this in collaboration or with your guidance :)

WDYT?

**Comment by vincent99** [comment]
There is never going to be a generic "Oauth for X".  As I said in the comment you linked to the basic redirect flow and token handling is already is a laughable non-standard with no two major implementations alike.  Beyond that there isn't even an attempt to standardize all the other bits we need to make it useful, like looking up users and groups.

The idea was to move all the backend logic for a service into an golang [auth service provider](https://github.com/rancher/rancher-auth-service/tree/master/providers) to make them less tied to the actual product and easier for us to add new ones.  Github was moved over as a POC and then Shibboleth was added as the first new one.  The others will be migrated eventually.

It is possible to write your own; we don't have any way to load 3rd party providers but it's just a binary in the container, or we could take PRs if it's for something generally applicable.  There is a common pattern with the UI redirecting to the auth service which then redirects to the provider, so it is probably possible to make the login integration generic easily for things that have no other arguments.  Configuring would be more complicated but it is all just JSON sent to the auth service and ultimately cattle settings in the API so you don't necessarily need configuration UI.

**Comment by chicco785** [comment]
Hi @vincent99, 
we agree that there is not generic "Oauth for X". Maybe it was not clear, but what we would like to generalise or se generalised, is the way new identity / authentication providers can be integrated without having the pain of rebuilding multiple rancher services :) (at least this was the case for the rancher version @taliaga and @Cerfoglg played with).

Ideally, all the configuration complexity should be in "auth service provider" that you can add dynamically (like you do for example with the machine-drivers) keeping the interface "generic" and agnostic to single implementations. Configuration itself can either be integrated in the driver, or, as you state, pushed through API Calls.

Said so, we are happy to help on that, if this considered relevant to the community. In which case we are happy to follow your guidelines to implement the needed features, or propose our self the design, depending on how you feel more comfortable :)

**Comment by vincent99** [comment]
1.12 is not a Rancher version so it's not clear if you meant `1.1.2` (which was before the auth service existed) or `1.2` (after).  Maybe you can give more detail about what you changed because minus the loading dynamically part that's basically already how it should be; all the logic for Shibboleth is in `rancher-auth-service`, plus login and configuration in the UI.

**Comment by Cerfoglg** [comment]
@vincent99 What we've tried to do, very roughly, is replace the Github provider implementation in the Auth Service with our own custom implementation to work with a different Oauth provider. We altered the Auth Service, built the dist for it, then rebuilt a new rancher server image by first rebuilding the cattle distribution including our altered auth service, then building the server image and including the custom cattle dist.

This way we were able to get a Rancher server with access control using our Oauth provider, with configuration passed via the Rancher API, working more or less as intended. Our dilemma now is how to easily maintain our custom Rancher server image, including keeping up with the official updates to Rancher. I would assume that as long as we can maintain our custom provider implementation to be used in the Auth Service then it should not be too much hassle if we wanted to rebuild the custom image with newer versions of Rancher.

Feel free to correct me if I've misunderstood how the server image is meant to be built, and if this process can be simplified.
